<a href="https://colab.research.google.com/github/barkle2/Programming/blob/master/%5BPython%5D%20Tutorial/kosis_API_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 통계청 OPEN API 활용 예제

### 1. 인증키 받기
- 가장 먼저 해야하는 일은 KOSIS 공유서비스(https://kosis.kr/openapi/index/index.jsp) 사이트에 가입하고, 인증키를 받는 것입니다.
- 인증키 유효기간은 2년입니다.

In [1]:
key = "MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ="

### 2. 통계목록 확인
- 요청 URL: http://kosis.kr/openapi/statisticsList.do?method=getList
- 요청 변수(String)
  - apiKey : 인증키
  - vwCd
    - MT_ZTITLE : 국내통계 주제별
    - MT_OTITLE : 국내통계 기관별
  - parentListId
    - MT_ZTITLE : 노동(D)
    - MT_OTITLE : 통계청(101), 고용노동부(118)
  - format: json

- API 호출을 위한 Query 문 생성

In [2]:
query = "https://kosis.kr/openapi/statisticsList.do?method=getList&apiKey={}&vwCd={}&parentListId={}&format=json&jsonVD=Y"
vwCd = "MT_ZTITLE"
parentListId = "D"

query = query.format(key, vwCd, parentListId)

- requests library를 활용하여 API 호출

In [3]:
import requests
import pandas as pd
from google.colab import files
import time, random

- '국내통계 주제별' 중 '노동'에 해당하는 통계목록은 다음과 같다.

In [4]:
r = requests.get(query)
r.content
data = r.json()
df = pd.DataFrame(data)
df

LIST_NM      VW_CD     VW_NM LIST_ID
0   경력단절여성등의경제활동실태조사  MT_ZTITLE  국내통계 주제별     D_1
1           경제활동인구조사  MT_ZTITLE  국내통계 주제별     D_2
2            고령자고용현황  MT_ZTITLE  국내통계 주제별     D_3
3             고용보험통계  MT_ZTITLE  국내통계 주제별     D_4
4          고용허가제고용동향  MT_ZTITLE  국내통계 주제별     D_5
5        고용형태별근로실태조사  MT_ZTITLE  국내통계 주제별     D_6
6           구직급여신청동향  MT_ZTITLE  국내통계 주제별     D_7
7             근로환경조사  MT_ZTITLE  국내통계 주제별     D_8
8         기업직업훈련실태조사  MT_ZTITLE  국내통계 주제별     D_9
9       기업체장애인고용실태조사  MT_ZTITLE  국내통계 주제별    D_10
10           노동생산성지수  MT_ZTITLE  국내통계 주제별    D_11
11            노사분규통계  MT_ZTITLE  국내통계 주제별    D_12
12       대졸자직업이동경로조사  MT_ZTITLE  국내통계 주제별    D_13
13          박사인력활동조사  MT_ZTITLE  국내통계 주제별    D_14
14     사업체기간제근로자현황조사  MT_ZTITLE  국내통계 주제별    D_15
15          사업체노동력조사  MT_ZTITLE  국내통계 주제별    D_16
16         사업체노동실태현황  MT_ZTITLE  국내통계 주제별    D_17
17      산업기술인력수급실태조사  MT_ZTITLE  국내통계 주제별    D_18
18            산재보험통계  MT_ZTITLE  국내통계 주제별    D_19
19          산재보험패널조사  MT_ZTITLE  국내통계 주제별    D_20
20         여성관리자패널조사  MT_ZTITLE  국내통계 주제별    D_21
21      워크넷구인구직및취업동향  MT_ZTITLE  국내통계 주제별    D_22
22      이민자체류실태및고용조사  MT_ZTITLE  국내통계 주제별    D_23
23           일자리이동통계  MT_ZTITLE  국내통계 주제별   D_001
24           일자리행정통계  MT_ZTITLE  국내통계 주제별    D_24
25     임금근로일자리동향행정통계  MT_ZTITLE  국내통계 주제별    D_25
26            육아휴직통계  MT_ZTITLE  국내통계 주제별   B_001
27          작업환경실태조사  MT_ZTITLE  국내통계 주제별    D_26
28       장애인경제활동실태조사  MT_ZTITLE  국내통계 주제별    D_27
29         장애인고용패널조사  MT_ZTITLE  국내통계 주제별    D_28
30      장애인구인구직및취업동향  MT_ZTITLE  국내통계 주제별    D_29
31         장애인의무고용현황  MT_ZTITLE  국내통계 주제별    D_30
32        전국노동조합조직현황  MT_ZTITLE  국내통계 주제별    D_31
33           지역별고용조사  MT_ZTITLE  국내통계 주제별    D_32
34       직종별사업체노동력조사  MT_ZTITLE  국내통계 주제별    D_33
35            청년패널조사  MT_ZTITLE  국내통계 주제별    D_34
36          한국노동패널조사  MT_ZTITLE  국내통계 주제별    D_35

- 경제활동인구조사 통계 내에도 여러 개의 세부 통계목록이 존재한다.

In [5]:
query = "https://kosis.kr/openapi/statisticsList.do?method=getList&apiKey={}&vwCd={}&parentListId={}&format=json&jsonVD=Y"
vwCd = "MT_ZTITLE"
parentListId = "D_2"

query = query.format(key, vwCd, parentListId)

In [6]:
r = requests.get(query)
r.content
data = r.json()
df = pd.DataFrame(data)
df

LIST_NM  ...  LIST_ID
0             경제활동인구총괄(공식 실업률, 구직기간 4주기준, 1999.6~현재)  ...      B11
1                             계절조정 경제활동인구(1999.6~현재)  ...      B12
2                                       취업자(1963~현재)  ...      B17
3    실업자(구직기간 4주 기준, 1999.5월 이전자료는 경제활동인구총괄 1주기준 참조)  ...      B16
4   실업률(구직기간 4주 기준, 1999. 5월 이전자료는 경제활동인구총괄 1주기준 참조)  ...      B15
5           비경제활동인구(1999. 5월 이전자료는 경제활동인구총괄 1주기준 참조)  ...      B14
6                  경제활동인구총괄(구직기간 1주기준, 1963~2014.12)  ...      B13
7                                         근로형태별 부가조사  ...  101_B1A
8                                           청년층 부가조사  ...   B1A_18
9                                           고령층 부가조사  ...   B1A_20
10                                        비임금근로 부가조사  ...   B1A_22
11                                비임금근로 부가조사(농림어업제외)  ...   B1A_24
12                                      비경제활동인구 부가조사  ...   B1A_26
13                         시계열 보정(2018.2.14 공표) 前 자료  ...  B1A_101

[14 rows x 4 columns]

- 가장 많이 사용되는 '경제활동인구총괄(공식 실업률, 구직기간 4주기준, 1999.6~현재)' 표의 세부 목록을 확인한다.

In [7]:
query = "https://kosis.kr/openapi/statisticsList.do?method=getList&apiKey={}&vwCd={}&parentListId={}&format=json&jsonVD=Y"
vwCd = "MT_ZTITLE"
parentListId = "B11"

query = query.format(key, vwCd, parentListId)

In [8]:
r = requests.get(query)
r.content
data = r.json()
df = pd.DataFrame(data)
df

TBL_NM        TBL_ID      VW_CD     VW_NM ORG_ID
0            성별 경제활동인구 총괄   DT_1DA7001S  MT_ZTITLE  국내통계 주제별    101
1           연령별 경제활동인구 총괄   DT_1DA7002S  MT_ZTITLE  국내통계 주제별    101
2            교육정도별 경제활동인구   DT_1DA7003S  MT_ZTITLE  국내통계 주제별    101
3        행정구역(시도)별 경제활동인구   DT_1DA7004S  MT_ZTITLE  국내통계 주제별    101
4            성/연령별 경제활동인구   DT_1DA7012S  MT_ZTITLE  국내통계 주제별    101
5          성/교육정도별 경제활동인구   DT_1DA7013S  MT_ZTITLE  국내통계 주제별    101
6      행정구역(시도)/성별 경제활동인구   DT_1DA7014S  MT_ZTITLE  국내통계 주제별    101
7     행정구역(시도)/연령별 경제활동인구   DT_1DA7015S  MT_ZTITLE  국내통계 주제별    101
8   행정구역(시도)/교육정도별 경제활동인구   DT_1DA7016S  MT_ZTITLE  국내통계 주제별    101
9              고용보조지표(전체)   DT_1DA7300S  MT_ZTITLE  국내통계 주제별    101
10    고용보조지표(청년층, 15~29세)  DT_1DA7300AS  MT_ZTITLE  국내통계 주제별    101

- '연령별 경제활동인구 총괄' 표의 통계자료를 조회하기 위해 필요한 테이블 ID(TBL_ID:DT_1DA7002S)와 기관 ID(ORG_ID:101)를 얻었다.

In [9]:
tbl_id = 'DT_1DA7012S'
org_id = '101'

## 3. 통계자료 조회

### 3-1. 최신 시점 찾기
- 현재 기준으로 가장 최신 통계가 언제의 통계인지 확인한다.
- '횡단면 조회' 자료를 받아 최신 통계 시점을 확인한다.
- 통계청 OPEN API 사이트에서 횡단면 조회 query 예제 복사

In [15]:
# 0:apiKey, 1:itmId, 2:objL1, 3:objL2, 4:prdSe, 5:loadGubun, 6:ordId, 7:tblId
query = "https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey={0}&itmId={1}+&objL1={2}+&objL2={3}&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe={4}&newEstPrdCnt=1&loadGubun={5}&orgId={6}&tblId={7}"

- 통계청 OPEN API 사이트에서 항목, 분류 value 찾기

In [16]:
# 항목
itm_id = {
    "인구":"T10",
    "경제활동인구":"T20",
    "취업자":"T30",
    "실업자":"T40",
    "비경제활동인구":"T50",
    "경제활동참가율":"T60",
    "실업률":"T80",
    "고용률":"T90"
}

# 성별 분류
sex_id = {
  "계": "0",
  "남자": "2",
  "여자": "3"
}

# 연령 분류
age_id = {
     "15세 이상 전체": "00",
     "15 - 19세": "10",
     "20 - 29세": "20",
     "20 - 24세": "201",
     "25 - 29세": "202",
     "30 - 39세": "30",
     "30 - 34세": "301",
     "35 - 39세": "302",
     "40 - 49세": "40",
     "40 - 44세": "401",
     "45 - 49세": "402",
     "50 - 59세": "50",
     "50 - 54세": "501",
     "55 - 59세": "502",
     "60세 이상": "60",
     "60 - 64세": "601",
     "65세 이상": "602",
     "15 - 64세": "63",
     "15 - 24세": "70",
     "15 - 29세": "75"
}



- 그 밖에 필요한 변수는 직접 입력

In [17]:
prd_se = 'M'
load_gubun = '2'

query = query.format(key, itm_id['인구'], sex_id['계'], age_id['15세 이상 전체'], prd_se, load_gubun, org_id, tbl_id)
query

'https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&itmId=T10+&objL1=0+&objL2=00&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe=M&newEstPrdCnt=1&loadGubun=2&orgId=101&tblId=DT_1DA7012S'

- '횡단면 조회'의 PRD_DE 값이 가장 최근 시점(end_prd_de)이 된다.

In [18]:
r = requests.get(query)
r.content
data = r.json()
df = pd.DataFrame(data)
df

TBL_NM  PRD_DE       TBL_ID   ITM_NM  ... C1_NM C2_NM C1_NM_ENG C2_NM_ENG
0  성/연령별 경제활동인구  202011  DT_1DA7012S  15세이상인구  ...     계     계     Total     Total

[1 rows x 21 columns]

In [19]:
end_prd_de = df['PRD_DE'][0]
end_prd_de

'202011'

- 가장 최근 시점을 구했으니 시계열 조회를 한다.
- 통계청 OPEN API 사이트에서 시계열 조회 query 예제 복사

In [20]:
# 0:apiKey, 1:itmId, 2:objL1, 3:objL2, 4:prdSe, 5:startPrdDe, 6:endPrdDe, 7:loadGubun, 8:orgId, 9:tblId
query = "https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey={0}&itmId={1}+&objL1={2}+&objL2={3}&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe={4}&startPrdDe={5}&endPrdDe={6}&loadGubun={7}&orgId={8}&tblId={9}"

start_prd_de = '199906'
load_gubun = '1'
org_id
tbl_id

query.format(key, itm_id['인구'], sex_id['계'], age_id['15세 이상 전체'], prd_se, start_prd_de, end_prd_de, load_gubun, org_id, tbl_id )

'https://kosis.kr/openapi/Param/statisticsParameterData.do?method=getList&apiKey=MDUwYjkzYmRhZjc4Yzk5M2QyZDg1NDBhNjNlOTNjZmQ=&itmId=T10+&objL1=0+&objL2=00&objL3=&objL4=&objL5=&objL6=&objL7=&objL8=&format=json&jsonVD=Y&prdSe=M&startPrdDe=199906&endPrdDe=202011&loadGubun=1&orgId=101&tblId=DT_1DA7012S'

In [21]:
# 자료 저장을 위한 DataFrame
df = pd.DataFrame()

- itm_id(항목)과 obj_lev2(연령구분)을 반복문으로 바꿔가며 조회하고, df에 모두 합친다.

In [22]:
for item in itm_id.values():
  for sex in sex_id.values():
    for obj in age_id.values():    
      query_temp = query.format(key, item, sex, obj, prd_se, start_prd_de, end_prd_de, load_gubun, org_id, tbl_id )        
      r = requests.get(query_temp)

      data = r.json()
      df_temp = pd.DataFrame(data)
      df = df.append(df_temp, ignore_index=True)
      print(item, sex, obj, df.shape)
      time.sleep(random.uniform(2,8))

T10 0 00 (258, 21)
T10 0 10 (516, 21)
T10 0 20 (774, 21)
T10 0 201 (1032, 21)
T10 0 202 (1290, 21)
T10 0 30 (1548, 21)
T10 0 301 (1806, 21)
T10 0 302 (2064, 21)
T10 0 40 (2322, 21)
T10 0 401 (2580, 21)
T10 0 402 (2838, 21)
T10 0 50 (3096, 21)
T10 0 501 (3354, 21)
T10 0 502 (3612, 21)
T10 0 60 (3870, 21)
T10 0 601 (4128, 21)
T10 0 602 (4386, 21)
T10 0 63 (4644, 21)
T10 0 70 (4902, 21)
T10 0 75 (5160, 21)
T10 2 00 (5418, 21)
T10 2 10 (5676, 21)
T10 2 20 (5934, 21)
T10 2 201 (6192, 21)
T10 2 202 (6450, 21)
T10 2 30 (6708, 21)
T10 2 301 (6966, 21)
T10 2 302 (7224, 21)
T10 2 40 (7482, 21)
T10 2 401 (7740, 21)
T10 2 402 (7998, 21)
T10 2 50 (8256, 21)
T10 2 501 (8514, 21)
T10 2 502 (8772, 21)
T10 2 60 (9030, 21)
T10 2 601 (9288, 21)
T10 2 602 (9546, 21)
T10 2 63 (9804, 21)
T10 2 70 (10062, 21)
T10 2 75 (10320, 21)
T10 3 00 (10578, 21)
T10 3 10 (10836, 21)
T10 3 20 (11094, 21)
T10 3 201 (11352, 21)
T10 3 202 (11610, 21)
T10 3 30 (11868, 21)
T10 3 301 (12126, 21)
T10 3 302 (12384, 21)
T10 3 40 

- 합쳐진 결과를 csv 파일로 저장한다.(euc-kr 또는 utf-8-sig 추천)

In [23]:
df.to_csv('경제활동인구조사(euc-kr).csv', encoding='euc-kr')
files.download('경제활동인구조사(euc-kr).csv')
df.to_csv('경제활동인구조사(utf-8).csv', encoding='utf-8')
files.download('경제활동인구조사(utf-8).csv')
df.to_csv('경제활동인구조사(utf-8-sig).csv', encoding='utf-8-sig')
files.download('경제활동인구조사(utf-8-sig).csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>